### PR TITLE
riscv: simplify state management during examine

### PR DESCRIPTION
**Pull request by aap-sc**
This also fixes a bug when, after `examine` completion, the target still
has  `unknown` status. To reproduce this one spike, it is enough to do
the following:

~~~
// make sure spike harts are halted
openocd ... -c init -c 'echo "[targets]"'
~~~

this behavior is quite dangerous and leads to segfaults in some cases

Change-Id: I13915f7038ad6d0251d56d2d519fbad9a2f13c18

### CODE REVIEW

File: src/target/riscv/riscv-013.c [hunk 1]
```diff
@@ -1972,20 +1975,9 @@ static int examine(struct target *target)
 		target->state = TARGET_HALTED;
 		target->debug_reason = DBG_REASON_DBGRQ;
```
Comment (MarekVCodasip) [inline src/target/riscv/riscv-013.c:1976]: I feel like this can be removed now as well.

Comment (MarekVCodasip) [inline src/target/riscv/riscv-013.c]: Seeing as the target-> state is set here and also in riscv013_halt_go and debug_reason later. I would just delete these two lines.

**Review by MarekVCodasip — COMMENTED** [review]
I think this is an improvement, especially later when the state is restored.

File: src/target/riscv/riscv-013.c (hunk 1 above)
Comment (JanMatCodasip) [reply src/target/riscv/riscv-013.c]: Agreed, this seems redundant. The change of the recorded target state will be covered by lines 1979 and 1980.

File: src/target/riscv/riscv-013.c [hunk 2]
```diff
@@ -1958,11 +1958,14 @@ static int examine(struct target *target)
...
-	if (riscv_get_hart_state(target, &state) != ERROR_OK)
+	target->state = TARGET_UNKNOWN;
+	target->debug_reason = DBG_REASON_UNDEFINED;
+
+	enum riscv_hart_state state_at_examine_start;
+	if (riscv_get_hart_state(target, &state_at_examine_start) != ERROR_OK)
 		return ERROR_FAIL;
-	bool halted = (state == RISCV_STATE_HALTED);
-	if (!halted) {
+	const bool hart_halted_at_examine_start = state_at_examine_start == RISCV_STATE_HALTED;
```
Comment (JanMatCodasip) [inline src/target/riscv/riscv-013.c:2043]: Just a side note: This part of code seems to not handle the case when the hart is unavailable (RISCV_STATE_UNAVAILABLE) - and neither did the the original code. 

No need to solve this in the current merge request. Just a remark for a future improvement.

**Review by JanMatCodasip — DISMISSED** [review]
This looks all right to me.

File: src/target/riscv/riscv-013.c [hunk 3]
```diff
@@ -1958,11 +1958,14 @@ static int examine(struct target *target)
 	if (dm013_select_hart(target, info->index) != ERROR_OK)
 		return ERROR_FAIL;
 
-	enum riscv_hart_state state;
-	if (riscv_get_hart_state(target, &state) != ERROR_OK)
+	target->state = TARGET_UNKNOWN;
+	target->debug_reason = DBG_REASON_UNDEFINED;
```
Comment (timsifive) [inline src/target/riscv/riscv-013.c]: What does this achieve? I don't think it's accessed below, it's just unconditionally overwritten.

File: src/target/riscv/riscv-013.c (hunk 1 above)
Comment (timsifive) [reply src/target/riscv/riscv-013.c]: Agreed. Remove these 2 lines.

File: src/target/riscv/riscv-013.c [hunk 4]
```diff
@@ -2059,13 +2051,13 @@ static int examine(struct target *target)
...
-	target->debug_reason = saved_dbg_reason;
-
-	if (!halted) {
+	if (state_at_examine_start == RISCV_STATE_RUNNING) {
 		riscv013_step_or_resume_current_hart(target, false);
 		target->state = TARGET_RUNNING;
 		target->debug_reason = DBG_REASON_NOTHALTED;
+	} else if (state_at_examine_start == RISCV_STATE_HALTED) {
+		target->state = TARGET_HALTED;
+		target->debug_reason = DBG_REASON_UNDEFINED;
```
Comment (timsifive) [inline src/target/riscv/riscv-013.c:2134]: Why is this necessary? We know that the hart is halted on line 2054, and the state should reflect that.

**Review by timsifive — CHANGES_REQUESTED** [review]

File: src/target/riscv/riscv-013.c (hunk 4 above)
Comment (EccoTheDolphin) [reply src/target/riscv/riscv-013.c:2134]: > Why is this necessary? We know that the hart is halted on line 2054, and the state should reflect that.

Yes, we know that the hart is halted. And we have `target->state = TARGET_HALTED` to indicate that. However, we don't know **why** hart was halted (since we've just connected) and thus we set the reason to UNDEFINED.

File: src/target/riscv/riscv-013.c (hunk 3 above)
Comment (EccoTheDolphin) [reply src/target/riscv/riscv-013.c]: `What does this achieve? I don't think it's accessed below, it's just unconditionally overwritten.`

This is for the case when examine fails. Imagine we could not examine/could not finish examination process. This allows us just to reset the state from whatever it was. My understanding is that examine (regardless of the current implementation) can actually be invoked several times (via `arp_examine` for example). See here:

https://sourceforge.net/p/openocd/mailman/message/33948884/

File: src/target/riscv/riscv-013.c (hunk 2 above)
Comment (EccoTheDolphin) [reply src/target/riscv/riscv-013.c:2043]: Well... actually it does. Kind of. Upon entry to examine routine we unconditionally set:

```
	target->state = TARGET_UNKNOWN;
	target->debug_reason = DBG_REASON_UNDEFINED;
```

And then the assumption is that if the hart is UNAVAILABLE - then all operations should fail (like halt_go).

File: src/target/riscv/riscv-013.c (hunk 1 above)
Comment (EccoTheDolphin) [reply src/target/riscv/riscv-013.c]: I'll take look. on the first glance, yes it looks like we should remove this.

File: src/target/riscv/riscv-013.c (hunk 3 above)
Comment (EccoTheDolphin) [reply src/target/riscv/riscv-013.c]: corrected the message. `arp_init` -> `arp_examine`

File: src/target/riscv/riscv-013.c (hunk 2 above)
Comment (JanMatCodasip) [reply src/target/riscv/riscv-013.c:2043]: Ok, thank you for explanation. (No need to change anything at the moment.)

File: src/target/riscv/riscv-013.c (hunk 1 above)
Comment (aap-sc) [reply src/target/riscv/riscv-013.c]: addressed

**Review by JanMatCodasip — DISMISSED** [review]

File: src/target/riscv/riscv-013.c (hunk 3 above)
Comment (timsifive) [reply src/target/riscv/riscv-013.c]: OK. Can you add a short comment, like `/* In case e.g. riscv013_halt_go() fails. */`

Generally I would like to see these status updates happen where we know the status changed, but that's a much larger change.

File: src/target/riscv/riscv-013.c (hunk 4 above)
Comment (timsifive) [reply src/target/riscv/riscv-013.c:2134]: Got it. Thanks.

**Review by timsifive — CHANGES_REQUESTED** [review]
Just a minor change request. Otherwise good to go.

File: src/target/riscv/riscv-013.c (hunk 3 above)
Comment (EccoTheDolphin) [reply src/target/riscv/riscv-013.c]: > OK. Can you add a short comment, like /* In case e.g. riscv013_halt_go() fails. */

Addressed. Please do note that I moved these to lines way higher. I think this is more appropriate place.

**Review by timsifive — APPROVED** [review]

**Review by JanMatCodasip — APPROVED** [review]
